### PR TITLE
Add `DictionaryArray::key` function

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?
Closes https://github.com/apache/arrow-rs/issues/1911

# Rationale for this change
See https://github.com/apache/arrow-rs/issues/1911
 
# What changes are included in this PR?

1. Add `DictionaryArray::key` function
2. Tests

# Are there any user-facing changes?

Yes new function

### CODE REVIEW

File: arrow/src/array/array_dictionary.rs [hunk 1]
```diff
@@ -169,6 +169,17 @@ impl<'a, K: ArrowPrimitiveType> DictionaryArray<K> {
...
     }
+
+    /// Return the value of `keys` (the dictionary key) at index `i`,
+    /// cast to `usize`, `None` if the value at `i` is `NULL`.
+    pub fn key(&self, i: usize) -> Option<usize> {
+        self.keys.is_valid(i).then(|| {
+            self.keys
+                .value(i)
+                .to_usize()
+                .expect("Dictionary index not usize")
```
Comment (waynexia) [inline arrow/src/array/array_dictionary.rs:180]: For most cases this unwrap won't panic, but do we need to maintain the same behavior with https://github.com/apache/arrow-datafusion/blob/080c32400ddfa2d45b5bebb820184eac8fd5a03a/datafusion/common/src/scalar.rs#L342-L358 ? If so the return type can either be `Result<Option<_>>` or `Option<_>`, I'm ok with both (hard to choose...).

**Review by waynexia — APPROVED** [review]
This is a nice interface :+1: I left a comment about that `expect`.

**Review by viirya — APPROVED** [review]
Looks good to me. Currently to I need to call `value` on `keys` array. This should be convenient.

File: arrow/src/array/array_dictionary.rs (hunk 1 above)
Comment (tustvold) [reply arrow/src/array/array_dictionary.rs:180]: Panicking should be fine, it's a validation failure if the array contains negative indexes. TBH I keep meaning to change all these checked conversions to numeric casts (i.e. as), I wouldn't be surprised if this leads to non-trivial performance benefits.

**Review by tustvold — APPROVED** [review]

**Pull request merged**